farm-staking/farm-staking/src/lib.rs

```diff
@@ -96,19 +97,83 @@ pub trait FarmStaking:
         let current_epoch = self.blockchain().get_block_epoch();
         self.first_week_start_epoch().set_if_empty(current_epoch);
 
+        // Initialize last_reward_timestamp
+        let current_timestamp = self.blockchain().get_block_timestamp();
+        self.last_reward_timestamp().set_if_empty(current_timestamp);
+
         // Farm position migration code
         let farm_token_mapper = self.farm_token();
         self.try_set_farm_position_migration_nonce(farm_token_mapper);
     }
 
     #[upgrade]
     fn upgrade(&self) {
-        let current_epoch = self.blockchain().get_block_epoch();
-        self.first_week_start_epoch().set_if_empty(current_epoch);
+        let mut storage_cache = StorageCache::new(self);
 
-        // Farm position migration code
-        let farm_token_mapper = self.farm_token();
-        self.try_set_farm_position_migration_nonce(farm_token_mapper);
+        // GENERATE AGGREGATED REWARDS
+        let accumulated_rewards_mapper = self.accumulated_rewards();
+        let mut accumulated_rewards = accumulated_rewards_mapper.get();
+        let reward_capacity = self.reward_capacity().get();
+        let remaining_rewards = &reward_capacity - &accumulated_rewards;
+
+        // MINT PER BLOCK REWARDS
+        let last_reward_block_nonce_mapper =
+            SingleValueMapper::<Self::Api, u64>::new(StorageKey::new(b"last_reward_block_nonce"));
+        let per_block_reward_amount_mapper = SingleValueMapper::<Self::Api, BigUint>::new(
+            StorageKey::new(b"per_block_reward_amount"),
+        );
+
+        let current_block_nonce = self.blockchain().get_block_nonce();
+        let last_reward_nonce = last_reward_block_nonce_mapper.take();
+        let per_block_reward = per_block_reward_amount_mapper.take();
+
+        require!(
+            self.produces_per_second_rewards(),
+            "Farm must produce rewards"
+        );
+
+        // CALCULATE PER BLOCK REWARDS
+        let extra_rewards_unbounded = if current_block_nonce <= last_reward_nonce {
+            BigUint::zero()
+        } else {
+            let block_nonce_diff = current_block_nonce - last_reward_nonce;
+            &per_block_reward * block_nonce_diff
+        };
+
+        let farm_token_supply = self.farm_token_supply().get();
+        let max_apr = self.max_annual_percentage_rewards().get();
+        let extra_rewards_apr_bounded_per_block =
+            farm_token_supply * &max_apr / MAX_PERCENT / (SECONDS_IN_YEAR / 6u64);
+
+        let block_nonce_diff = current_block_nonce - last_reward_nonce;
```

Comment:
You can move this line above the `extra_rewards_unbounded` and use inside the if statement